apps/omg/lib/omg/state/transaction/payment.ex

```diff
@@ -0,0 +1,324 @@
+  alias OMG.Utxo
+
+  require Transaction
+  require Utxo
+  require Transaction.Payment
+
+  @zero_address OMG.Eth.zero_address()
+  @empty_signature <<0::size(520)>>
+
+  # TODO: commented code for the tx markers handling
+  # @payment_marker Transaction.Markers.payment()
+  #
+  # end commented code
+
+  @doc """
+  Turns a structure instance into a structure of RLP items, ready to be RLP encoded, for a raw transaction
+  """
+  def get_data_for_rlp(%Transaction.Payment{inputs: inputs, outputs: outputs, metadata: metadata})
+      when Transaction.Payment.is_metadata(metadata),
+      do:
+        [
+          # TODO: commented code for the tx markers handling
+          # @payment_marker,
+          # contract expects 4 inputs and outputs
+          Enum.map(inputs, fn %{blknum: blknum, txindex: txindex, oindex: oindex} -> [blknum, txindex, oindex] end) ++
+            List.duplicate([0, 0, 0], 4 - length(inputs)),
+          Enum.map(outputs, fn %{owner: owner, currency: currency, amount: amount} -> [owner, currency, amount] end) ++
+            List.duplicate([@zero_address, @zero_address, 0], 4 - length(outputs))
+        ] ++ if(metadata, do: [metadata], else: [])
+
+  def get_outputs(%Transaction.Payment{outputs: outputs}) do
+    outputs
+    |> Enum.reject(&match?(%{owner: @zero_address, currency: @zero_address, amount: 0}, &1))
+  end
+
+  def get_inputs(%Transaction.Payment{inputs: inputs}) do
+    inputs
+    |> Enum.map(fn %{blknum: blknum, txindex: txindex, oindex: oindex} -> Utxo.position(blknum, txindex, oindex) end)
+    |> Enum.filter(&Utxo.Position.non_zero?/1)
+  end
+
+  def valid?(%Transaction.Payment{}, %Transaction.Signed{sigs: sigs} = tx) do
+    tx
+    |> Transaction.get_inputs()
+    |> all_inputs_signed?(sigs)
+  end
+
+  @doc """
+  True if a payment can be applied, given a set of input UTXOs is present in the ledger.
+  Involves the checking of balancing of inputs and outputs for currencies
+
+  Returns the fees that this transaction is paying, mapped by currency
+  """
+  @spec can_apply?(Transaction.Payment.t(), list(Utxo.t())) :: {:ok, map()} | {:error, :amounts_do_not_add_up}
+  def can_apply?(%Transaction.Payment{} = tx, input_utxos) do
+    outputs = Transaction.get_outputs(tx)
+
+    input_amounts_by_currency = get_amounts_by_currency(input_utxos)
+    output_amounts_by_currency = get_amounts_by_currency(outputs)
+
+    with :ok <- amounts_add_up?(input_amounts_by_currency, output_amounts_by_currency),
+         do: {:ok, fees_paid(input_amounts_by_currency, output_amounts_by_currency)}
+  end
+
+  @doc """
+  Effects of a payment transaction - spends all inputs and creates all outputs
+  """
+  def get_effects(%Transaction.Payment{} = tx, blknum, tx_index) do
+    new_utxos_map = tx |> non_zero_utxos_from(blknum, tx_index) |> Map.new()
+    spent_input_pointers = Transaction.get_inputs(tx)
+    {spent_input_pointers, new_utxos_map}
+  end
+
+  defp all_inputs_signed?(non_zero_inputs, sigs) do
```

Comment:
a couple of points to reply to that:

> I have a feeling it's not good place for this fn. Also does it bring any value? I think authorized? already check this, kind of.

but `authorized?` is stateful validity

> Does it make difference that tx is missing a sig or is invalidly signed?

it kinda depends on whether our transactions will still require that (pre-stateful-validity checks about the coherence of witnesses delivered). So we had this (the fixed requirement, that all inputs are signed, according to their position), and this was executed in stateless validity. In order for the implementations to be equivalent, `Transaction.Protocol.valid?` needed to pop up :(. 

But this also is bugging me a bit. Maybe if this turns out to be an unnecessary abstraction (we'll be returning `true` always anyway) we can remove.

> Also how about Transaction.Signed how it fit in proposed model. Wouldn't be better to have this fn there?

Unfortunately not, because the presence and behavior of this particular check is transaction-type dependent.

Whew, this is a bit convoluted, but I hope this makes sense